### PR TITLE
 add functional test for bridge dp 

### PR DESCRIPTION
**Pull request by phoracek**
With this test we try to deploy the plugin, attach Pods to provided resources and ping between them.

Run with:

```
make cluster-up build-network-bridge cluster-sync-network-bridge functest
```

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -11,3 +14,20 @@ def disable_insecure_requests_warning():
 @pytest.fixture(scope="session", autouse=True)
 def load_local_cluster_kubeconfig():
     config.load_kube_config('./kubeconfig')
+
+
+@pytest.fixture
+def temporary_namespace():
```
Comment (EdDev) [inline tests/conftest.py]: You can call it namespace or k8s_namespace. The fact that it is temporary is embedded in the fixture itself.

And do you really need it at a test level scope? Won't a session level be good? If session level, you can even use `autouse` to avoid mentioning it as a dependency in the test modules.

File: tests/kubeutils.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+from kubernetes.stream import stream
+
+
+RESOURCE_READY = 0
+RESOURCE_EMPTY = 1
+RESOURCE_MISSING = 2
+
+
+def assert_resource_availability(resource, node, status, timeout=0):
+    v1 = client.CoreV1Api()
```
Comment (EdDev) [inline tests/kubeutils.py]: Can it be a module/session fixture or just a global?

File: tests/kubeutils.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    w = watch.Watch()
+    for event in w.stream(v1.list_node, timeout_seconds=timeout):
+        if event['object'].metadata.name == node:
+            if (status == RESOURCE_READY and
+                    resource in event['object'].status.allocatable and
```
Comment (EdDev) [inline tests/kubeutils.py]: event['object'].status.allocatable.get(resource, '0') != '0'

File: tests/kubeutils.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                    event['object'].status.allocatable[resource] != '0'):
+                return
+            elif (status == RESOURCE_EMPTY and
+                    resource in event['object'].status.allocatable and
```
Comment (EdDev) [inline tests/kubeutils.py]: event['object'].status.allocatable.get(resource) == '0'

File: tests/kubeutils.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+from kubernetes.client import rest
+from kubernetes.stream import stream
+
+
+RESOURCE_READY = 0
+RESOURCE_EMPTY = 1
+RESOURCE_MISSING = 2
+
+
+def assert_resource_availability(resource, node, status, timeout=0):
```
Comment (EdDev) [inline tests/kubeutils.py]: If it is the caller who sets the status (explicitly), then it will be more readable to use 3 different asserts, one for each status, something lie: assert_resource_ready()

File: tests/kubeutils.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    v1 = client.CoreV1Api()
+
+    w = watch.Watch()
+    for event in w.stream(v1.list_node, timeout_seconds=timeout):
```
Comment (EdDev) [inline tests/kubeutils.py]: How do you know if the event has not been already dispatched before calling this?

Comment (phoracek) [reply tests/kubeutils.py]: It first lists all existing as "ADDED".

File: tests/conftest.py (hunk 1 above)
Comment (phoracek) [reply tests/conftest.py]: I don't want to handle created pods etc cleanup myself. I'd keep it this way and change it once we have more tests.

File: tests/kubeutils.py (hunk 2 above)
Comment (phoracek) [reply tests/kubeutils.py]: I would need to pass the fixture everywhere, to all helper functions. I don't want to do that.

I will try to make it global, but I think it broke it the last time.

File: tests/kubeutils.py (hunk 5 above)
Comment (phoracek) [reply tests/kubeutils.py]: Good point, done.

File: tests/kubeutils.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+from kubernetes import client
+from kubernetes import watch
+from kubernetes.client import rest
+from kubernetes.stream import stream
+
+
+def assert_resource_ready(resource, node, timeout=0):
+    _assert_resource_status(
+        node=node,
+        status_check=lambda resources: resources.get(resource, '0') != '0',
```
Comment (EdDev) [inline tests/kubeutils.py:12]: You could make it simpler and easier to read:
resource_check=lambda resource: resource != '0'

The "resources.get(resource, '0')" is common, so it can stay in the backend function.

File: tests/kubeutils.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+def _assert_resource_status(node, status_check, timeout=0):
+    v1 = client.CoreV1Api()
+
+    w = watch.Watch()
+    for event in w.stream(v1.list_node, timeout_seconds=timeout):
+        if event['object'].metadata.name == node:
+            if status_check(event['object'].status.allocatable):
+                return
+
+    raise AssertionError('Given resource have not turned into expected state')
```
Comment (EdDev) [inline tests/kubeutils.py:34]: Note that you mention a single resource, so you better use "has" in the message.
When it raises, do you have enough info to debug? (same question for the other assertions)

File: tests/kubeutils.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    for event in w.stream(v1.list_namespaced_pod, namespace,
+                          timeout_seconds=timeout):
+        if (event['object'].metadata.name and
+                event['object'].status.phase == state):
+            return
+
+    raise AssertionError('Pods have not turned into the expected state')
+
+
+def run(pod, namespace, command):
```
Comment (EdDev) [inline tests/kubeutils.py:65]: Is this for local usage only? If it is indeed public, then:
I would recommend to split assertions from other function helpers, but you could do that in a separate PR/commit. You do not want these mixed, it will mess up the module.

It will also be nice to add context to "what" are you running exactly.

File: tests/kubeutils.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            pod, namespace, _preload_content=False)
+        try:
+            assert expected_string in logs.read()
+        except AssertionError:
+            if i != timeout:
+                time.sleep(1)
+            else:
+                raise
+        else:
+            return
```
Comment (EdDev) [inline tests/kubeutils.py:90]: How about locating this "return" in the try body?
There are other cases like this below.

**Review by EdDev — COMMENTED** [review]
Partial review

File: tests/netutils.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+RESOURCES_VARIABLE_NAME_PREFIX = 'NETWORK_INTERFACE_RESOURCES_'
+
+
+def read_interfaces_by_resource(pod, namespace):
+    interfaces_by_resource = collections.defaultdict(list)
+
+    env_raw = kubeutils.run(pod, namespace, 'env')
+    env = dict([pair.split('=', 1)
+                for pair in [line for line in env_raw.split('\n')[:-1]]])
```
Comment (EdDev) [inline tests/netutils.py:15]: Why do you need the "pair"? Isn't this enough: line for line in env_raw.split('\n')[:-1]
If you want it for readability, just use a variable to store the lines/pairs.

Also, why do you need a dict? A tuple seems to be sufficient here.

File: tests/netutils.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+    for name, value_raw in env.items():
+        if name.startswith(RESOURCES_VARIABLE_NAME_PREFIX):
+            value = json.loads(value_raw)
+            interfaces_by_resource[value['name']].extend(value['interfaces'])
```
Comment (EdDev) [inline tests/netutils.py:20]: How about using '+' instead of extend?

I'm not exactly clear when you use the default value of [] for interfaces_by_resource (it is a defaultdict).

File: tests/network_bridge_test.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+import yaml
+
+from kubernetes import client
+
+from . import kubeutils
+from . import netutils
+from . import nodes
+
+
+def test_bridge_device_plugin(namespace):
```
Comment (EdDev) [inline tests/network_bridge_test.py:12]: Looks like this is testing too many things and it makes things hard to understand. If something will fail, there will be a need to examine all this huge test.

Please try to split it a bit, locating the setup stuff outside in fixtures/calsses and make the tests small.

File: tests/network_bridge_test.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+    with open('./tests/network_bridge_test_pods.yml') as f:
+        consumer_pods = yaml.load(f)
+
+    v1 = client.CoreV1Api()
+    apps_v1 = client.AppsV1Api()
+
+    # Create a config map listing names of bridges that will be exposed as
+    # available resources
+    v1.create_namespaced_config_map(
```
Comment (EdDev) [inline tests/network_bridge_test.py:27]: This and other like it are oneliners.

File: tests/network_bridge_test.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+
+    # Wait until the daemon set is ready on all nodes
+    kubeutils.assert_daemon_set_ready(
+        daemon_set=device_plugin_daemon_set['metadata']['name'],
+        namespace=namespace,
+        timeout=300
+    )
+
+    # Verify that all device plugin's instances started successfully
+    for pod in kubeutils.list_daemon_set_pods(
```
Comment (EdDev) [inline tests/network_bridge_test.py:46]: Please separate and use a variable, it should make it nicer.

File: tests/network_bridge_test.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    for pod in consumer_pods['items']:
+        kubeutils.assert_pod_state(
+            pod['metadata']['name'],
+            namespace=namespace,
+            state='Pending',
+            timeout=300
+        )
+
+    # Configure networks across cluster nodes
+    with _configure_host_networking('node01'), \
```
Comment (EdDev) [inline tests/network_bridge_test.py:88]: If you call it "_configure_host_network" it will fit one line.

File: tests/network_bridge_test.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+@contextmanager
+def _configure_host_networking(node):
+    # On KubeVirt CI nodes, forwarding is disabled by default, enable it
+    nodes.ssh(node, ['sudo', 'iptables', '--policy', 'FORWARD', 'ACCEPT'])
+
+    # Create two VLAN interfaces on top of eth0 and connect them to separate
+    # bridges. That will give us two isolated networks between nodes
+    for net_id in [1, 2]:
+        nodes.ssh(node, """
```
Comment (EdDev) [inline tests/network_bridge_test.py:154]: Please do not define the command inline like this.

File: tests/network_bridge_test.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+sudo ip link add link eth0 name eth0.{net_id} type vlan id {net_id} &&
+sudo ip link add name mybr{net_id} type bridge &&
+sudo ip link set eth0.{net_id} master mybr{net_id} &&
+sudo ip link set eth0.{net_id} up &&
+sudo ip link set mybr{net_id} up
+        """.format(net_id=net_id).split())
+    try:
+        yield
+    finally:
```
Comment (EdDev) [inline tests/network_bridge_test.py:163]: This is not safe, you may have leftovers if part of the command above fail before even entering the try block.
